storage-kit/src/p2p/actor.rs

```diff
@@ -0,0 +1,122 @@
+use anyhow::{anyhow, Result};
+use tokio::{select, sync::{mpsc::Receiver, oneshot}};
+use std::time::Duration;
+use libp2p::{futures::StreamExt, identify, identity::Keypair, noise, swarm::NetworkBehaviour, Multiaddr, Swarm};
+use tracing::*;
+
+use crate::{Actor, ActorEventLoop, Port};
+
+/// Defines max_negotiating_inbound_streams constant for the swarm.
+/// It must be set for large plots.
+const SWARM_MAX_NEGOTIATING_INBOUND_STREAMS: usize = 100000;
+
+/// How long will connection be allowed to be open without any usage.
+const IDLE_CONNECTION_TIMEOUT: Duration = Duration::from_secs(10);
+
+pub const IDENTIFY_PROTOCOL: &str = "/storagehub/id/0.0.1";
+
+#[derive(NetworkBehaviour)]
+pub(crate) struct Behaviour {
+	identify: identify::Behaviour,
+}
+
+pub struct P2PModule {
+    pub(crate) swarm: Swarm<Behaviour>,
+}
+
+impl P2PModule {
+	/// Creates a new swarm agent with the given identity on the given `port`.
+	pub fn new(identity: Keypair, port: Port) -> Result<P2PModule> {
+		let mut swarm = libp2p::SwarmBuilder::with_existing_identity(identity)
+			.with_tokio()
+			.with_tcp(
+				libp2p::tcp::Config::default(),
+				noise::Config::new,
+				libp2p::yamux::Config::default,
+			)?
+			.with_quic()
+			.with_behaviour(|key| {
+				Ok(Behaviour {
+					identify: identify::Behaviour::new(identify::Config::new(
+						IDENTIFY_PROTOCOL.into(),
+						key.public(),
+					)),
+				})
+			})?
+			.with_swarm_config(|config| {
+				config
+					.with_max_negotiating_inbound_streams(SWARM_MAX_NEGOTIATING_INBOUND_STREAMS)
+					.with_idle_connection_timeout(IDLE_CONNECTION_TIMEOUT)
+			})
+			.build();
+
+		// Listen on all interfaces on the specified port.
+		swarm.listen_on(format!("/ip4/0.0.0.0/tcp/{}", port).parse()?)?;
+
+		Ok(P2PModule {
+			swarm,
+		})
+	}
+}
+
+pub struct P2PEventLoop {
```

Comment:
Shouldn’t the `P2PEventLoop` have a hashmap of senders to send messages to other actors? 

Example:
```rust
senders: HashMap<ActorHandle<A>, Sender<C>>, // A being an ActorHandle and C being a Command type for an actor handle
```
This would give flexibility for the event loops like P2P or Blockchain module to send commands to each other but also give possibility to add other actors in the future.

Reply:
If there is no other way around, we can add it later.

However, I'd like to aim to have no cycles (or as little as possible) between actors since it's really easy to make deadlocks otherwise.

I don't see why the Actors have to talk directly with each other. I imagine the higher level tasks would "resolve" that.

Reply:
Also, this PR does not include yet the event dispatcher.

Each of the actors/modules will have access to a global event dispatcher (mostly to dispatch events). But there is no constraint why actor A couldn't register to actor B's events.

Reply:
Ok I see what you mean. But I don’t understand how the higher level process would not rely on the Blockchain module to begin a storage request for example. The higher level process manages all executions in some sequence using commands. Do you intend to have some higher level processes listen on blockchain events coming from the blockchain module to begin a process? 

1. Higher level process listening to events <- *__does it listen to events, and from where?__*
2. Receives storage request <- *__how does this happen__*
3. Spawn task (or loop listener) blockchain module
a. Execute volunteer transaction
b. Return success to higher level process
4. Send command (storage request) P2P module
a. Connect to user and get file
b. Return success to higher level process
5. Successfully processed storage request

Reply:
I've updated and implemented event busses in the meantime. Please check [storage-kit/src/tasks.rs](https://github.com/Moonsong-Labs/storage-hub/pull/21/files#diff-d0773164cc29978634c85691ef7dff6355c29c9a4117633ba5764f9b7635603d) to see how an event-triggered task would interact with the different actors/modules.
Check [this](https://github.com/Moonsong-Labs/storage-hub/pull/21/files#diff-68316775a6d5e33dfad7aff44fe5488f64547456625fb81e649a1a3aad9de422R22) to see how this task is registered.

Let me know if this is not clear enough or if it doesn't cover everything you were thinking of.